### ISSUE TITLE
More advanced expression editor

### ISSUE THREAD

**Issue by orangemug** [maputnik/editor#622]
Now we're getting close to the completion of adding basic expression support to the editor (see #223), here are my plans for more advanced editing. The approach is to add inline widgets into the JSON editor. A simple example is color

<img width="371" alt="Screenshot 2020-03-09 at 15 40 06" src="https://user-images.githubusercontent.com/235915/76232852-37544800-621f-11ea-84b7-3f12a8b26952.png">
<img width="370" alt="Screenshot 2020-03-09 at 16 14 01" src="https://user-images.githubusercontent.com/235915/76234031-0412b880-6221-11ea-8169-3a10b481ec5f.png">

There is a demo of that here <https://2354-84182601-gh.circle-artifacts.com/0/artifacts/build/index.html> - note it just displays colors with a click handler right now, open 'Zoomstack roads' and go to the 'sites' layer, to see what's in the screenshot.

Looking through the spec I believe we could add support for

 - Color - via a color picker, this would be for any value that matches a color string regardless of spec.
 - Number - from spec with min/max values defined
 - Image - pick image ids from spritesheet, would pick up `["image", "foo"]` in expressions.
 - Enum - Any field spec that has a enum

I think color is probably the most important here because colors are the most tricky thing to get right in a design. 

Anyway what do people think?

**Comment by nspringer-trimble** [comment]
Picking colors interactively and seeing the changes in real time is THE key feature to Maputnik for me as a map designer.  Being able to see subtle differences in color as you drag is necessary.

**Comment by orangemug** [comment]
Yeah agree @nspringer-trimble we'll definitely start with color. Is there anything else you think could be improved about the JSON Layer/Expression editors?

**Comment by nspringer-trimble** [comment]
Some kind of visual Scratch-like representation could still be nice: https://github.com/maputnik/editor/issues/223#issuecomment-411404757
Also validation/help would be good.

**Comment by orangemug** [comment]
@pathmapper I've removed this from v1.8.0 milestone, I think we should get the accessibility/deployment changes live first.

**Comment by orangemug** [comment]
@pathmapper @kylebarron what if this was the expression editor?

https://user-images.githubusercontent.com/235915/145889310-3c432dcd-8f13-4654-9027-9c6b25ef08b3.mov

https://user-images.githubusercontent.com/235915/145889400-1f8771dc-2437-43c9-9880-85b0429cb4e4.mov

**Comment by pathmapper** [comment]
@orangemug looks good from the first impression!

Would it be possible somehow to read the whole description from the style spec? Currently the description is cropped and ends with `...`

![grafik](https://user-images.githubusercontent.com/20856381/146051517-81ed4ab0-9cc9-46cb-96f7-086fd8d17523.png)

From where are the available properties with description are coming from? Is there need to have special metadata properties in the data? What happens if this (meta-)data is not available? 

![grafik](https://user-images.githubusercontent.com/20856381/146052326-77f39d23-24f9-404a-aea9-ee424b937e2a.png)

**Comment by orangemug** [comment]
> Would it be possible somehow to read the whole description from the style spec? Currently the description is cropped and ends with 

Yeah it all comes directly from the style spec so we could render it elsewhere also. So you can do `1+1` instead of `+(1, 1)` so it has the concept of binary expressions and math precedence. All the function name are slightly altered so `line_progress` rather than `line-progress`. Because the result of the expression is stored in JSON arrays, once you reload your expression the editor has to reformat it. But that's actually quite nice, because it allows for an auto-formatter so there is an possibility for context aware formatting on the syntax. An example is `interpolate(linear(), get("mag"), 1, 2, 8, 10)` which naturally wants to be formatted as argument pairs so

```
interpolate(
  linear(), get("mag"),
  1, 2,
  8, 10,
)
```

> From where are the available properties with description are coming from? Is there need to have special metadata properties in the data? What happens if this (meta-)data is not available?

The metadata is a bit "magic", I think it'll always have to be to be honestly. It's much like the rest of the app, when we have metadata we show it when we don't we just display nothing. This is currently hard coded for the demo but we can def get the properties names from some vector tiles and GeoJSON.

You can have a play around at https://orangemug.github.io/texprl/demo/maplibre/

The idea is a general purpose language(if you can call it that) and editor that can be used with `["operation", args...]` type runtimes, like maplibre expressions. There is a demo using a vector math runtime over at https://orangemug.github.io/texprl/demo/vector/ also. There are definitely some bugs right now 😁
